src/drivers/dw/uart.c

```diff
@@ -0,0 +1,266 @@
+	if (dw_uart_ring_tail == dw_uart_ring_head) {
+		dw_uart_ring_empty = true;
+		wait_completed(&dev->complete);
+	} else {
+		uart_write(dev, SUE_UART_REG_IER, IER_PTIME | IER_ETBEI);
+	}
+
+unlock:
+	spin_unlock(&dw_uart_ring_lock);
+}
+
+void dw_uart_init(void)
+{
+	struct dw_uart_device_full *dev = &uart_dev;
+	unsigned int irq;
+
+	/* first call - allocate the ring */
+	dw_uart_ring = rmalloc(RZONE_SYS, SOF_MEM_CAPS_RAM,
+			       DW_UART_RING_SIZE);
+	if (!dw_uart_ring)
+		return;
+
+	wait_init(&dev->complete);
+	spinlock_init(&dw_uart_ring_lock);
+
+	/* Register interrupt for the same core, where we're running */
+	irq = dev->irq | (cpu_get_id() << SOF_IRQ_CPU_SHIFT);
+	if (!interrupt_register(irq, IRQ_AUTO_UNMASK, dw_uart_irq_handler, dev))
+		interrupt_enable(irq);
+}
+
+void dw_uart_write_word(uint32_t word)
+{
+	uint32_t flags;
+
+	spin_lock_irq(&dw_uart_ring_lock, flags);
+	dw_uart_write_word_internal(&uart_dev.common, word);
+	spin_unlock_irq(&dw_uart_ring_lock, flags);
+}
+
+static void dw_uart_wait(void)
+{
+	struct dw_uart_device_full * const dev = &uart_dev;
+	int ret;
+
+	/* 100ms should be enough */
+	dev->complete.timeout = 100000;
+	ret = wait_for_completion_timeout(&dev->complete);
+	if (ret < 0) {
+		/* Timeout... */
+		dw_uart_write_word(0x6f000000);
+		dw_uart_write_word(uart_read(dev, SUE_UART_REG_IIR));
+	}
+}
+
+int dw_uart_write_nowait(const char *data, uint32_t size)
+{
+	struct dw_uart_device_full * const dev = &uart_dev;
+	unsigned int tail_room, head_room, count;
+	uint32_t flags;
+
+	if (!dw_uart_ring)
+		/* No buffer, abort */
+		return -ENOBUFS;
+
+	spin_lock_irq(&dw_uart_ring_lock, flags);
+
+	if (size <= 0 ||
+	    (!dw_uart_ring_empty &&
+	     dw_uart_ring_tail == dw_uart_ring_head)) {
+		/* No data or ring full */
+		spin_unlock_irq(&dw_uart_ring_lock, flags);
+		return size;
+	}
+
+	if (dw_uart_ring_tail <= dw_uart_ring_head) {
+		head_room = DW_UART_RING_SIZE - dw_uart_ring_head;
+		tail_room = dw_uart_ring_tail;
+	} else {
+		head_room = dw_uart_ring_tail - dw_uart_ring_head;
+		tail_room = 0;
+	}
+
+	count = min(head_room, size);
+	arch_memcpy(dw_uart_ring + dw_uart_ring_head, data, count);
+
+	dw_uart_ring_head += count;
+	if (dw_uart_ring_head == DW_UART_RING_SIZE)
+		dw_uart_ring_head = 0;
+
+	size -= count;
+
+	if (size && tail_room) {
+		count = min(tail_room, size);
+		arch_memcpy(dw_uart_ring, data + head_room, count);
+		dw_uart_ring_head = count;
+		size -= count;
+	}
+
+	dw_uart_ring_empty = false;
+
+	wait_clear(&dev->complete);
+
+	/* enable the TX empty interrupt */
+	uart_write(dev, SUE_UART_REG_IER, IER_PTIME | IER_ETBEI);
+
+	spin_unlock_irq(&dw_uart_ring_lock, flags);
+
+	return size;
+}
+
+/* Block until all the data is at least in the ring buffer */
+int dw_uart_write(const char *data, uint32_t size)
+{
+	for (;;) {
+		int ret = dw_uart_write_nowait(data, size);
+		if (ret <= 0)
+			return ret;
+
+		data += size - ret;
+		size = ret;
+
+		dw_uart_wait();
+	}
+
+	return 0;
+}
```

Comment:
I'd expect to see a struct uart_ops here with pointer to each UART operation. That way clients would call uart_write() instead of dw_write(). A generic struct sof_uart would also need to be defined and passed into each operation.

Reply:
can we also do it like I've done for SPI / GPIO / IOMUX? Just use generic names for exported functions and rely on linking no more than just one required driver? So I'll just drop "dw_" prefixes from exported functions and types.

Reply:
no, your solution breaks when > 1 devices exists. please use existing convention.

Reply:
You mean you want to be able to handle multiple UART models / drivers simultaneously? Ok, sure, this version only works when there's a single UART, used only for logging (the driver cannot even receive data). I'll add ops then.

Reply:
Yes exactly.